include/deal.II/base/thread_local_storage.h

```diff
@@ -193,40 +179,59 @@ namespace Threads
   // ----------------- inline and template functions --------------------------
 
   template <typename T>
-  inline ThreadLocalStorage<T>::ThreadLocalStorage(const T &t)
-    : data(t)
+  inline ThreadLocalStorage<T>::ThreadLocalStorage()
   {}
 
 
   template <typename T>
-  inline ThreadLocalStorage<T>::ThreadLocalStorage(
-    const ThreadLocalStorage<T> &t)
-    : data(t)
-  {}
+  inline ThreadLocalStorage<T>::ThreadLocalStorage(const T &t)
+  {
+    exemplar = std::make_shared<T>(t);
+  }
 
 
   template <typename T>
   inline T &
-  ThreadLocalStorage<T>::get()
+  ThreadLocalStorage<T>::get(bool &exists)
   {
-#    ifdef DEAL_II_WITH_THREADS
-    return data.local();
-#    else
-    return data;
-#    endif
+    const std::thread::id my_id = std::this_thread::get_id();
+
+    auto it = data.find(my_id);
+    if (it != data.end())
```

Comment:
You're accessing `data` here. You need to move the lock guard above this point.

Reply:
Also, this is most easily written as
```
  if (const auto it = data.find(...))
```
and then you don't need to declare the local variable here.

Reply:
This is actually the perfect example of a reader-writer lock. Shall we bump to C++14 in order to use `std::shared_lock` and `std::unique_lock`?

Reply:
Right, there's probably a design pattern for that, but the code as is is still wrong ;-)